include/gmsh.tcc

```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
       << "The coordinates for vertices of each element have been stored\n";
 }
 
-//! Compute material points based on the centroid
+//! Compute material points for 2D quadrilateral
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
-//! \param[in] ngauss_points Number of gauss points per coordinate
-template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::compute_material_points(unsigned ngauss_points) {
+//! \param[in] ngauss_points Number of gauss points per coordinate, element_type GMSH element type
+template <>
+inline void GMSH<2, 4>::compute_material_points(unsigned ngauss_points, unsigned element_type) {
+
+  const unsigned Tdim = 2;
+  const unsigned Tvertices = 4;
+
+  //! Storing ngauss_points to member variable and get constants from namespace
+  ngauss_points_ = ngauss_points;
+  std::vector<double> gauss_constants =
+      element::gauss_points.find(ngauss_points_)->second;
+
+  //! Create a matrix of xi from gauss points
+  //! Matrix is size npoints x Tdim
+  //! For 2D only
+  unsigned npoints = std::pow(ngauss_points_, Tdim);
+  Eigen::MatrixXd xi_gauss_points(npoints, Tdim);
+  unsigned counter = 0;
+  for (unsigned ii = 0; ii < ngauss_points_; ++ii) {
+    for (unsigned jj = 0; jj < ngauss_points_; ++jj) {
+        xi_gauss_points(counter, 0) = gauss_constants.at(ii);
+        xi_gauss_points(counter, 1) = gauss_constants.at(jj);
+        ++counter;
+    }
+  }
+
+  Eigen::VectorXd pointsarray(Tdim);
+
+  //! global_id is the index for all points
+  unsigned global_id = 0;
+  //! material_id is the index of materialpoints
+  unsigned material_id = 0;
+  //! point_id is local index of points
+  unsigned point_id = 0;
+
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::shared_ptr<MaterialPoints<Tdim>>(
+      new MaterialPoints<Tdim>(material_id)));
+  for (const auto& element : elements_) {
+
+    //! Store coordinates in Tdim x Tvertices matrix
+    //! Where N is the number of nodes per element
+    //! This is rearranging of the data to have stored in matrix form
+    Eigen::MatrixXd node_coordinates(Tdim, Tvertices);
```

Comment:
Do we need to store this matrix? Can we not call `vertex co-ordinates` directly in #269?

Reply:
I think we can, but to me, the loop after that would not be as elegant. I feel that we make it rather more elegant this way.

Reply:
OK